src/http/server.cr

```diff
@@ -95,72 +100,187 @@ class HTTP::Server
+  end
+
+  # Creates a new HTTP server with the given block as handler.
+  def self.new(&handler : HTTP::Handler::Proc) : self
+    new(handler)
+  end
+
+  # Creates a new HTTP server with a handler chain constructed from the *handlers*
+  # array and the given block.
+  def self.new(handlers : Array(HTTP::Handler), &handler : HTTP::Handler::Proc) : self
+    new(HTTP::Server.build_middleware(handlers, handler))
+  end
+
+  # Creates a new HTTP server with the *handlers* array as handler chain.
+  def self.new(handlers : Array(HTTP::Handler)) : self
+    new(HTTP::Server.build_middleware(handlers))
   end
 
-  def initialize(@host : String, @port : Int32, handler : HTTP::Handler | HTTP::Handler::Proc)
+  # Creates a new HTTP server with the given *handler*.
+  def initialize(handler : HTTP::Handler | HTTP::Handler::Proc)
     @processor = RequestProcessor.new(handler)
   end
 
-  # Returns the TCP port the server is connected to.
+  # Creates a `TCPServer` and adds it as an interface.
+  #
+  # If *port* is `0`, a random, free port will be chosen.
+  #
+  # You may set *reuse_port* to `true` to enable the `SO_REUSEPORT` socket option,
+  # which allows multiple processes to bind to the same port.
+  def bind(host : String, port : Int32, reuse_port : Bool = false) : TCPServer
+    bind TCPServer.new(host, port, reuse_port: reuse_port)
+  end
+
+  # Creates a `TCPServer` listenting on `127.0.0.1` and adds it as an interface.
+  #
+  # If *port* is `0`, a random, free port will be chosen.
+  #
+  # You may set *reuse_port* to true to enable the `SO_REUSEPORT` socket option,
+  # which allows multiple processes to bind to the same port.
+  def bind(port : Int32, reuse_port : Bool = false) : TCPServer
+    bind "127.0.0.1", port, reuse_port
+  end
+
+  # Creates a `TCPServer` listening on an unused port and adds it as an interface.
+  #
+  # Returns the determined port number.
+  #
+  # ```
+  # server = HTTP::Server.new { }
+  # server.bind_unused_port # => 12345
+  # ```
+  def bind_unused_port(host : String = "127.0.0.1", reuse_port : Bool = false) : Int32
+    tcp_server = bind(host, 0, reuse_port)
+    tcp_server.local_address.port
+  end
+
+  # Creates a `Socket::Server` listenting to *address* and adds it as an interface.
+  #
+  # The *address* can either be `host:port` or a Unix socket prefixed by `unix:`.
+  # This obviously works only on systems supporting Unix sockets.
   #
-  # For example you may let the system choose a port, then report it:
   # ```
-  # server = HTTP::Server.new(0) { }
-  # server.bind
-  # server.port # => 12345
+  # server = HTTP::Server.new { }
+  # server.bind "127.0.0.1:8080"
+  # server.bind "unix:/tmp/my-socket"
   # ```
-  def port
-    if server = @server
-      server.local_address.port.to_i
+  def bind(address : String, reuse_port : Bool = false) : Socket::Server
+    if address.starts_with?("unix:")
+      {% if flag?(:unix) %}
+      bind(Socket::UNIXAddress.new(address[5..-1])).as(Socket::Server)
```

Comment:
I'm not sure we want to continue spattering ifdefs around the codebase. If sockets aren't supported, we'll raise in `UNIXAddress.new` instead of everywhere it's used. Indent is wrong.

Reply:
Yeah, it's probably better this way. The whole Socket API looks like it needs a major refactoring anyway.

Reply:
Some duplication, here. Having different `#bind` overloads is already redundant, but introducing yet another one with a URI-like string, required for binding UNIX sockets, that feels too much and weirdly designed.

It should either be the only solution, or there should be explicitly named methods (ala puma), which IMHO feels better (an application may require a URI, parse it, and call the correct methods):

- `#bind_tcp_listener(host, port)`
- `#bind_ssl_listener(host, port, ctx)`
- `#bind_unix_listener(host, port)`

Reply:
The UNIX socket part is removed from this PR. Regarding method naming, I'd suggest to continue the discussion in the main thread.